README.md

[![Build Status](https://travis-ci.org/KazuCocoa/simple_app_reporter_ex.svg?branch=master)](https://travis-ci.org/KazuCocoa/simple_app_reporter_ex)

# About

`Reporter` is simple reporting App reviews library.
Support AppStore and GooglePlay.

Hex is [here](https://hex.pm/packages/reporter)

Doc is [here](http://hexdocs.pm/reporter/extra-api-reference.html)

## Installation

First, add Reporter to your `mix.exs` dependencies:

```elixir
def deps do
  [
    {:reporter, "~> 0.6.3"}
  ]
end
```

and run `$ mix deps.get`.

# Usage
### Getting reviews from AppStore
#### JSON

```elixir
# The following method provide us very useful summary.
# You can get reviewed date, author, rating, title and body as Map.
iex> Reporter.app_store_rss_json!("375380948", "jp") |> Reporter.AppStore.review_summaries |> Enum.at(0)
%{"author" => "m.aaa...", "rating" => "4", "title" => "あまり", "body" => "利便性がわかりずらい"}


iex> Reporter.app_store_rss_json!("375380948", "jp")
     |> Reporter.AppStore.author
%{"name" => %{"label" => "iTunes Store"},
  "uri" => %{"label" => "http://www.apple.com/jp/itunes/"}}

iex> Reporter.app_store_rss_json!("375380948", "jp") |> Reporter.AppStore.reviews |> Enum.at(0)
%{"author" => %{"label" => "", "name" => %{"label" => "m.aaa..."},
  "uri" => %{"label" => "https://itunes.apple.com/jp/reviews/id451145371"}},
  "content" => %{"attributes" => %{"type" => "text"},
  "label" => "利便性がわかりずらい"},
  "id" => %{"label" => "1205430410"},
  "im:contentType" => %{"attributes" => %{"label" => "アプリケーション",
  "term" => "Application"}}, "im:rating" => %{"label" => "4"},
  "im:version" => %{"label" => "3.3"}, "im:voteCount" => %{"label" => "0"},
  "im:voteSum" => %{"label" => "0"},
  "link" => %{"attributes" => %{"href" => "https://itunes.apple.com/jp/review?id=375380948&type=Purple%20Software",
  "rel" => "related"}}, "title" => %{"label" => "あまり"}}
```

```elixir
iex> Reporter.app_store_rss_json!("375380948", "jp")
     |> Reporter.AppStore.all_json

# Return full JSON.
```

#### XML

Only implemented getting parsed XML.

```elixir
iex> Reporter.app_store_rss_xml!("375380948", "jp")
     |> Reporter.AppStore.all_xml

# Return XML based on FeederEx.
```

### Getting reviews from GooglePlay

```elixir
# The following method provide us very useful summary.
# You can get reviewed date, author, rating, title and body as Map.
iex> Reporter.google_play!("com.android.chrome", "en") |> Reporter.GooglePlay.review_summaries |> Enum.at(0)
%{"author" => "Kinoshanth Kino",
  "author_link" => "https://play.google.com/store/people/details?id=108550484713070099642",
  "body" => " It's really nice ",
  "date" => "May 22, 2016",
  "id" => "gp:AOqpTOHxPPI_l4M7AQy0djAtdTgDlgJUASN_rzS9tSelR6bQfN4cwRtYFNqjzTzzJX_qJeTXwQkNW4Ucsn1eJFk",
  "permalink" => "https://play.google.com/store/apps/details?id=com.android.chrome&reviewId=Z3A6QU9xcFRPSHhQUElfbDRNN0FReTBkakF0ZFRnRGxnSlVBU05fcnpTOXRTZWxSNmJRZk40Y3dSdFlGTnFqelR6ekpYX3FKZVRYd1FrTlc0VWNzbjFlSkZr",
  "rating" => 5.0,
  "title" => "Good"}

iex> Reporter.google_play!("com.android.chrome", "en") |> Reporter.GooglePlay.review_body_list |> Enum.at(0)
" Crashing a lot lot lot😡😣 "

iex> Reporter.google_play!("com.android.chrome", "en") |> Reporter.GooglePlay.review_bodies |> Enum.at(1)
{"div", [{"class", "review-body"}],
 [{"span", [{"class", "review-title"}], ["Not bad"]},
  " Unable to download from chrome. ",
  {"div", [{"class", "review-link"}, {"style", "display:none"}],
   [{"a",
     [{"class", "id-no-nav play-button tiny"}, {"href", "#"},
      {"target", "_blank"}], [" Full Review "]}]}]}
```

# Customize app store url and google play one

Generate `config/config.exs` and put the followings.

```
use Mix.Config

config :reporter,
  app_host: "https://itunes.apple.com",
  play_host: "https://play.google.com"
```

# LICENSE
license is MIT


CHANGELOG.md
# Changelog
## v0.7.0: Jul 23, 2017
- update dependencies
- Support Elixir 1.4+

## v0.6.3
- add 400 error handling

## v0.6.2
  - fix google play's author name is blank case

## v0.6.1
- use `certifi` 0.4.0 to avoid ssl error

## v0.6.0
- Support over Elixir 1.1.0

## v0.5.4
- Use `URI` to generate URL

## v0.5.3
- update dependencies

## v0.5.2
- Fix some warning ofElixir 1.3

## v0.5.1
- Fix parsing some XML escaped charactors
    - &, ", ', <, >

## v0.5.0
- add attributes of `author_link` and `permalink` in `Reporter.GooglePlay.review_summaries/1`
    - Please read readme if you'd like to know more.

### Breaking change
- drop `droid_uri` from `config/config.exs` and introduce `play_host` instead.
    - BTW, you don't need to set `play_host` becase "https://play.google.com" is used by default.

## v0.4.3
- add page_num in `google_play!/3`
- return id in addition from google play

## v0.4.2
- update deps

## 0.4.1
- fix wrong request parameters to GooglePlay.

## 0.4.0
- update dependencies
    - Use Poison 2.0.

## 0.3.3
- update dependencies

## 0.3.2
- update deps

## 0.3.1
- update dependencies

## ~ 0.3.0
- read release tags


test/reporter_test.exs
defmodule ReporterTest do
  use ExUnit.Case, async: true
  doctest Reporter.AppStore
  doctest Reporter.GooglePlay
end


test/reporter/app_store_test.exs
defmodule Reporter.AppStoreTest do
  use ExUnit.Case, async: true
  doctest Reporter.AppStore

  alias Reporter.AppStore

  setup_all do
    json = File.stream!("./test/data/ios_review.json", [], :line) |> Enum.into("") |> Poison.decode!
    json_empty = File.stream!("./test/data/ios_review_empty.json") |> Enum.into("") |> Poison.decode!
    {:ok, json: json, empty: json_empty}
  end

  test "AppStore.all_json", %{json: json, empty: json_empty} do
    assert AppStore.all_json(json) != nil
    assert AppStore.all_json(json_empty) != nil
  end

  test "AppStore.author", %{json: json, empty: json_empty} do
    assert AppStore.author(json) != nil
    assert AppStore.author(json_empty) != nil
  end

  test "AppStore.category", %{json: json, empty: json_empty} do
    category = AppStore.category(json)
    assert Enum.count(category) == 11
    assert AppStore.category(json_empty) == %{}
  end

  test "AppStore.reviews", %{json: json, empty: json_empty} do
    reviews = AppStore.reviews(json)
    assert Enum.count(reviews) == 50
    assert AppStore.reviews(json_empty) == %{}
  end

  test "AppStore.icon", %{json: json, empty: json_empty} do
    assert AppStore.icon(json) != nil
    assert AppStore.icon(json_empty) != nil
  end

  test "AppStore.id", %{json: json, empty: json_empty} do
    assert AppStore.id(json) != nil
    assert AppStore.id(json_empty) != nil
  end

  test "AppStore.link", %{json: json, empty: json_empty} do
    assert AppStore.link(json) != nil
    assert AppStore.link(json_empty) != nil
  end

  test "AppStore.rights", %{json: json, empty: json_empty} do
    assert AppStore.rights(json) != nil
    assert AppStore.rights(json_empty) != nil
  end

  test "AppStore.title", %{json: json, empty: json_empty} do
    assert AppStore.title(json) != nil
    assert AppStore.title(json_empty) != nil
  end

  test "AppStore.updated", %{json: json, empty: json_empty} do
    assert AppStore.updated(json) != nil
    assert AppStore.updated(json_empty) != nil
  end

end


test/reporter/google_play_test.exs
defmodule Reporter.GooglePlayTest do
  use ExUnit.Case, async: true
  doctest Reporter.GooglePlay
end


mix.exs
defmodule Reporter.Mixfile do
  use Mix.Project

  @version File.read!("VERSION") |> String.trim

  def project do
    [
      app: :reporter,
      version: @version,
      elixir: "~> 1.4",
      name: "Reporter",
      source_url: "https://github.com/KazuCocoa/simple_app_reporter_ex",
      description: "Simple getting reviews library from AppStore and GooglePlay",
      build_embedded: Mix.env == :prod,
      start_permanent: Mix.env == :prod,
      deps: deps(),
      package: package()
     ]
  end

  def application do
    [
      extra_applications: []
    ]
  end

  defp deps do
    [
      {:httpoison, "~> 0.8"},
      {:poison, "~> 3.0"},
      {:feeder_ex, "~> 1.0", runtime: false},
      {:floki, "~> 0.8", runtime: false},
      {:earmark, "~> 1.0", only: :dev, runtime: false},
      {:ex_doc, "~> 0.11", only: :dev, runtime: false},
      {:dialyze, "~> 0.2.0", only: :dev, runtime: false}
    ]
  end

  defp package do
    [
      files: ~w(lib mix.exs README.md LICENSE VERSION),
      maintainers: ["Kazuaki Matsuo"],
      licenses: ["MIT"],
      links: %{"GitHub" => "https://github.com/KazuCocoa/simple_app_reporter_ex"}
    ]
  end
end


lib/reporter.ex
defmodule Reporter do
  @moduledoc """
  `Reporter` is simple reporting App reviews library.
  Support AppStore and GooglePlay.
  """

  alias Reporter.AppStore
  alias Reporter.GooglePlay

  @doc """
  Get JSON response decoded by Poison.decode! from Apple Server.

  error case
  ### Not found
  iex> error = %{"message" => "Not found items.", "status_code" => "404"}
  iex> error["status_code"]
  "404"

  """
  @spec app_store_rss_json!(String.t, String.t) :: String.t
  def app_store_rss_json!(app_id, locale \\ "en") do
    headers = [{"Accept", "application/json; charset=UTF-8"}]
    HTTPoison.get(AppStore.rss_json(app_id, locale), headers)
    |> get_body_json!
  end

  @spec app_store_rss_json(String.t, String.t) :: {:error, String.t} | {:ok, String.t}
  def app_store_rss_json(app_id, locale \\ "en") do
    try do
      result = app_store_rss_json!(app_id, locale)
      {:ok, result}
    rescue
      e -> {:error, e.message}
    end
  end

  defp get_body_json!({:ok, %HTTPoison.Response{status_code: 200, body: body}}), do: body |> Poison.decode!
  defp get_body_json!({:ok, %HTTPoison.Response{status_code: 400}}), do: error_400()
  defp get_body_json!({:ok, %HTTPoison.Response{status_code: 404}}), do: error_404()
  defp get_body_json!({:error, %HTTPoison.Error{reason: reason}}), do: Poison.decode!(reason)

  @doc """
  Get XML formatted response from Apple Server.
  """
  @spec app_store_rss_xml!(String.t, String.t) :: String.t
  def app_store_rss_xml!(app_id, locale \\ "en") do
    headers = [{"Accept", "application/xml; charset=UTF-8"}]
    HTTPoison.get(AppStore.rss_xml(app_id, locale), headers)
    |> get_body_xml!
  end

  @spec app_store_rss_xml(String.t, String.t) :: {:error, String.t} | {:ok, String.t}
  def app_store_rss_xml(app_id, locale \\ "en") do
    try do
      result = app_store_rss_xml!(app_id, locale)
      {:ok, result}
    rescue
      e -> {:error, e.message}
    end
  end

  defp get_body_xml!({:ok, %HTTPoison.Response{status_code: 200, body: body}}), do: body
  defp get_body_xml!({:ok, %HTTPoison.Response{status_code: 400}}), do: error_400()
  defp get_body_xml!({:ok, %HTTPoison.Response{status_code: 404}}), do: error_404()
  defp get_body_xml!({:error, %HTTPoison.Error{reason: reason}}), do: IO.inspect reason

  @doc """
  Get HTML formatted response from GooglePlay.

  ## Example

  error case
  ### Not found
  iex> error = %{"message" => "Not found items.", "status_code" => "404"}
  iex> error["status_code"]
  "404"

  """
  @spec google_play!(String.t, String.t, Integer.t) :: list | map() | String.t
  def google_play!(package, locale \\ "en", page_num \\ 0) do
    headers = [{"Accept", "text/html; charset=UTF-8"}]
    HTTPoison.post(GooglePlay.review_url_with_page(package, page_num, locale), "", headers)
    |> get_body!
  end

  @spec google_play(String.t, String.t, Integer.t) :: {:ok, list} | {:error, String.t}
  def google_play(package, locale \\ "en", page_num \\ 0) do
    try do
      result = google_play!(package, locale, page_num)
      {:ok, result}
    rescue
      e -> {:error, e.message}
    end
  end


  defp get_body!({:ok, %HTTPoison.Response{status_code: 200, body: body}}) do
    body
    |> String.replace("\\u003c", "<")
    |> String.replace("\\u003e", ">")
    |> String.replace("\\u003d", "=")
    |> String.replace(~r/\\/, "")
    |> Floki.parse
    |> Enum.drop(1)
  end
  defp get_body!({:ok, %HTTPoison.Response{status_code: 400}}), do: error_400()
  defp get_body!({:ok, %HTTPoison.Response{status_code: 404}}), do: error_404()
  defp get_body!({:error, %HTTPoison.Error{reason: reason}}), do: reason

  defp error_400 do
    Map.new
    |> Map.put("status_code", "400")
    |> Map.put("message", "Bad Request")
  end

  defp error_404 do
    Map.new
    |> Map.put("status_code", "404")
    |> Map.put("message", "Not found items.")
  end
end


lib/reporter/app_store.ex
defmodule Reporter.AppStore do

  defp appstore_host(), do: Application.get_env(:reporter, :app_host, "https://itunes.apple.com")

  @doc """
  Return full JSON map.
  """
  @spec all_json(String.t) :: String.t
  def all_json(json), do: json

  @doc """
  Return full XML via Quinn.parse(xml).
  """
  @spec all_xml(String.t) :: String.t
  def all_xml(xml) do
    {:ok, feeder, _} = FeederEx.parse(xml)
    feeder
  end

  @doc ~S"""
  Returns collection of reviews.

  ## Examples

      iex> File.stream!("./test/data/ios_review.json") |> Enum.into("") |> Poison.decode! |> Reporter.AppStore.author
      %{"name" => %{"label" => "iTunes Store"},
        "uri" => %{"label" => "http://www.apple.com/jp/itunes/"}}

      iex> File.stream!("./test/data/ios_review_empty.json") |> Enum.into("") |> Poison.decode! |> Reporter.AppStore.author
      %{"name" => %{"label" => "iTunes Store"},
        "uri" => %{"label" => "http://www.apple.com/itunes/"}}
  """
  @spec author(Parser.t) :: map()
  def author(json), do: json["feed"]["author"]

  @doc ~S"""
  Returns collection of entry.

  ## Examples

      iex> File.stream!("./test/data/ios_review.json") |> Enum.into("") |> Poison.decode! |> Reporter.AppStore.category
      %{"category" => %{"attributes" => %{"im:id" => "6012",
            "label" => "ライフスタイル",
            "scheme" => "https://itunes.apple.com/jp/genre/ios-raifusutairu/id6012?mt=8&uo=2",
            "term" => "Lifestyle"}},
        "id" => %{"attributes" => %{"im:bundleId" => "com.apple.store.Jolly",
            "im:id" => "375380948"},
          "label" => "https://itunes.apple.com/jp/app/apple-store/id375380948?mt=8&uo=2"},
        "im:artist" => %{"attributes" => %{"href" => "https://itunes.apple.com/jp/artist/apple/id284417353?mt=8&uo=2"},
          "label" => "Apple"},
        "im:contentType" => %{"attributes" => %{"label" => "アプリケーション",
            "term" => "Application"}},
        "im:image" => [%{"attributes" => %{"height" => "53"},
           "label" => "http://is5.mzstatic.com/image/pf/us/r30/Purple1/v4/82/bd/90/82bd90c7-f673-718c-7c32-ad9db7591ea6/mzl.hfuxogmw.53x53-50.png"},
         %{"attributes" => %{"height" => "75"},
           "label" => "http://is1.mzstatic.com/image/pf/us/r30/Purple1/v4/82/bd/90/82bd90c7-f673-718c-7c32-ad9db7591ea6/mzl.hfuxogmw.75x75-65.png"},
         %{"attributes" => %{"height" => "100"},
           "label" => "http://is3.mzstatic.com/image/pf/us/r30/Purple1/v4/82/bd/90/82bd90c7-f673-718c-7c32-ad9db7591ea6/mzl.hfuxogmw.100x100-75.png"}],
        "im:name" => %{"label" => "Apple Store"},
        "im:price" => %{"attributes" => %{"amount" => "0.00000", "currency" => "JPY"},
          "label" => "入手"},
        "im:releaseDate" => %{"attributes" => %{"label" => "2011年9月21日"},
          "label" => "2011-09-21T10:10:33-07:00"},
        "link" => %{"attributes" => %{"href" => "https://itunes.apple.com/jp/app/apple-store/id375380948?mt=8&uo=2",
            "rel" => "alternate", "type" => "text/html"}},
        "rights" => %{"label" => "© 2015 Apple Inc."},
        "title" => %{"label" => "Apple Store - Apple"}}

      iex> File.stream!("./test/data/ios_review_empty.json") |> Enum.into("") |> Poison.decode! |> Reporter.AppStore.category
      %{}
  """
  @spec category(Parser.t) :: map()
  def category(json) do
    case json["feed"]["entry"] do
      nil -> %{}
      category -> category |> Enum.at(0)
    end
  end

  @doc ~S"""
  Returns collection of reviews.

  ## Examples

      iex> File.stream!("./test/data/ios_review.json") |> Enum.into("") |> Poison.decode! |> Reporter.AppStore.reviews |> Enum.at(0)
      %{"author" => %{"label" => "", "name" => %{"label" => "m.aaa..."},
        "uri" => %{"label" => "https://itunes.apple.com/jp/reviews/id451145371"}},
        "content" => %{"attributes" => %{"type" => "text"},
        "label" => "利便性がわかりずらい"},
        "id" => %{"label" => "1205430410"},
        "im:contentType" => %{"attributes" => %{"label" => "アプリケーション",
        "term" => "Application"}}, "im:rating" => %{"label" => "4"},
        "im:version" => %{"label" => "3.3"}, "im:voteCount" => %{"label" => "0"},
        "im:voteSum" => %{"label" => "0"},
        "link" => %{"attributes" => %{"href" => "https://itunes.apple.com/jp/review?id=375380948&type=Purple%20Software",
        "rel" => "related"}}, "title" => %{"label" => "あまり"}}

      iex> File.stream!("./test/data/ios_review_empty.json") |> Enum.into("") |> Poison.decode! |> Reporter.AppStore.reviews
      %{}
  """
  @spec reviews(Parser.t) :: map() | [map()] | map()
  def reviews(json) do
    case json["feed"]["entry"] do
      nil -> %{}
      entry -> entry |> Enum.drop(1)
    end
  end


  @doc ~S"""
  Returns list of reviews which are summarized.

  ## Examples

      iex> File.stream!("./test/data/ios_review.json") |> Enum.into("") |> Poison.decode! |> Reporter.AppStore.review_summaries |> Enum.at(0)
      %{"author" => "m.aaa...", "rating" => "4", "title" => "あまり", "body" => "利便性がわかりずらい"}

      iex> File.stream!("./test/data/ios_review_empty.json") |> Enum.into("") |> Poison.decode! |> Reporter.AppStore.review_summaries
      []
  """
  @spec review_summaries(Parser.t) :: [map()] | []
  def review_summaries(json) do
    json
    |> reviews()
    |> Enum.reverse
    |> Enum.reduce([], fn (review, list) ->
      author = review["author"]["name"]["label"]
      rating = review["im:rating"]["label"]
      title = review["title"]["label"]
      body = review["content"]["label"]

      result = Map.new
               |> Map.put("author", author)
               |> Map.put("rating", rating)
               |> Map.put("title", title)
               |> Map.put("body", body)

      List.insert_at(list, 0, result)
    end)
  end


  @doc ~S"""
  Returns icon.

  ## Examples

      iex> File.stream!("./test/data/ios_review.json") |> Enum.into("") |> Poison.decode! |> Reporter.AppStore.icon
      %{"label" => "http://itunes.apple.com/favicon.ico"}


      iex> File.stream!("./test/data/ios_review_empty.json") |> Enum.into("") |> Poison.decode! |> Reporter.AppStore.icon
      %{"label" => "http://itunes.apple.com/favicon.ico"}
  """
  @spec icon(Parser.t) :: map()
  def icon(json), do: json["feed"]["icon"]

  @doc ~S"""
  Returns id.

  ## Examples

      iex> File.stream!("./test/data/ios_review.json") |> Enum.into("") |> Poison.decode! |> Reporter.AppStore.id
      %{"label" => "https://itunes.apple.com/jp/rss/customerreviews/id%3d375380948/sortby%3dmostrecent/json"}

      iex> File.stream!("./test/data/ios_review_empty.json") |> Enum.into("") |> Poison.decode! |> Reporter.AppStore.id
      %{"label" => "https://itunes.apple.com/en/rss/customerreviews/id%3d375380948/sortby%3dmostrecent/json"}
  """
  @spec id(Parser.t) :: map()
  def id(json), do: json["feed"]["id"]

  @doc ~S"""
  Returns collection of link.

  ## Examples

    iex> File.stream!("./test/data/ios_review.json") |> Enum.into("") |> Poison.decode! |> Reporter.AppStore.link
    [%{"attributes" => %{"href" => "https://itunes.apple.com/WebObjects/MZStore.woa/wa/viewGrouping?cc=jp&id=26105",
                    "rel" => "alternate", "type" => "text/html"}},
                %{"attributes" => %{"href" => "https://itunes.apple.com/jp/rss/customerreviews/id=375380948/sortby=mostrecent/json",
                    "rel" => "self"}},
                %{"attributes" => %{"href" => "https://itunes.apple.com/jp/rss/customerreviews/page=1/id=375380948/sortby=mostrecent/xml?urlDesc=/customerreviews/id=375380948/sortby=mostrecent/json",
                    "rel" => "first"}},
                %{"attributes" => %{"href" => "https://itunes.apple.com/jp/rss/customerreviews/page=8/id=375380948/sortby=mostrecent/xml?urlDesc=/customerreviews/id=375380948/sortby=mostrecent/json",
                    "rel" => "last"}},
                %{"attributes" => %{"href" => "https://itunes.apple.com/jp/rss/customerreviews/page=1/id=375380948/sortby=mostrecent/xml?urlDesc=/customerreviews/id=375380948/sortby=mostrecent/json",
                    "rel" => "previous"}},
                %{"attributes" => %{"href" => "https://itunes.apple.com/jp/rss/customerreviews/page=2/id=375380948/sortby=mostrecent/xml?urlDesc=/customerreviews/id=375380948/sortby=mostrecent/json",
                    "rel" => "next"}}]

    iex> File.stream!("./test/data/ios_review_empty.json") |> Enum.into("") |> Poison.decode! |> Reporter.AppStore.link
    [%{"attributes" => %{"href" => "https://itunes.apple.com/WebObjects/MZStore.woa/wa/viewGrouping?cc=us&id=38",
                    "rel" => "alternate", "type" => "text/html"}},
                %{"attributes" => %{"href" => "https://itunes.apple.com/en/rss/customerreviews/id%3d375380948/sortby%3dmostrecent/json",
                    "rel" => "self"}},
                %{"attributes" => %{"href" => "", "rel" => "first"}},
                %{"attributes" => %{"href" => "", "rel" => "last"}},
                %{"attributes" => %{"href" => "", "rel" => "previous"}},
                %{"attributes" => %{"href" => "", "rel" => "next"}}]


  """
  @spec link(Parser.t) :: [map()] | []
  def link(json), do: json["feed"]["link"]

  @doc ~S"""
  Returns rights.

  ## Examples

    iex> File.stream!("./test/data/ios_review.json") |> Enum.into("") |> Poison.decode! |> Reporter.AppStore.rights
    %{"label" => "Copyright 2008 Apple Inc."}

    iex> File.stream!("./test/data/ios_review_empty.json") |> Enum.into("") |> Poison.decode! |> Reporter.AppStore.rights
    %{"label" => "Copyright 2008 Apple Inc."}

  """
  @spec rights(Parser.t) :: map()
  def rights(json), do: json["feed"]["rights"]

  @doc ~S"""
  Returns collection of title.

  ## Examples

    iex> File.stream!("./test/data/ios_review.json") |> Enum.into("") |> Poison.decode! |> Reporter.AppStore.title
    %{"label" => "iTunes Store: カスタマーレビュー"}

    iex> File.stream!("./test/data/ios_review_empty.json") |> Enum.into("") |> Poison.decode! |> Reporter.AppStore.title
    %{"label" => "iTunes Store: Customer Reviews"}

  """
  @spec title(Parser.t) :: map()
  def title(json), do: json["feed"]["title"]

  @doc ~S"""
  Returns time of updated.

  ## Examples

    iex> File.stream!("./test/data/ios_review.json") |> Enum.into("") |> Poison.decode! |> Reporter.AppStore.updated
    %{"label" => "2015-06-18T07:17:49-07:00"}

    iex> File.stream!("./test/data/ios_review_empty.json") |> Enum.into("") |> Poison.decode! |> Reporter.AppStore.updated
    %{"label" => "2015-06-21T06:52:45-07:00"}

  """
  @spec updated(Parser.t) :: map()
  def updated(json), do: json["feed"]["updated"]

  @doc ~S"""
  Return URLs with JSON.

  ## Examples

    iex> Reporter.AppStore.rss_json("375380948")
    "https://itunes.apple.com/en/rss/customerreviews/id=375380948/sortby=mostrecent/json"

    iex> Reporter.AppStore.rss_json("375380948", "jp")
    "https://itunes.apple.com/jp/rss/customerreviews/id=375380948/sortby=mostrecent/json"
  """
  @spec rss_json(String.t, String.t) :: String.t
  def rss_json(app_id, locale \\ "en"), do: rss(app_id, locale, "json")

  @doc ~S"""
  Return URLs with XML.

  ## Examples

      iex> Reporter.AppStore.rss_xml("375380948")
      "https://itunes.apple.com/en/rss/customerreviews/id=375380948/sortby=mostrecent/xml"

      iex> Reporter.AppStore.rss_xml("375380948", "jp")
      "https://itunes.apple.com/jp/rss/customerreviews/id=375380948/sortby=mostrecent/xml"
  """
  @spec rss_xml(String.t, String.t) :: String.t
  def rss_xml(app_id, locale \\ "en"), do: rss(app_id, locale, "xml")

  defp rss(app_id, locale, format) do
    Enum.join([
      appstore_host(),
      "/",
      locale,
      "/rss/customerreviews/id=",
      app_id,
      "/sortby=mostrecent/",
      format
      ]
    ) |> URI.encode
  end
end
